test/t8_schemes/t8_gtest_element_ref_coords.cxx

```diff
@@ -0,0 +1,184 @@
+}
+
+void
+t8_get_batch_coords_for_element_type (const t8_element_shape_t shape, double *batch_coords)
+{
+  const int num_vertices = t8_eclass_num_vertices[shape];
+  const int elem_dim = t8_eclass_to_dimension[shape];
+  for (int i_vertex = 0; i_vertex < num_vertices; ++i_vertex) {
+    for (int dim = 0; dim < elem_dim; ++dim) {
+      batch_coords[i_vertex * elem_dim + dim] = t8_element_corner_ref_coords[shape][i_vertex][dim];
+    }
+  }
+}
+
+template <typename T>
+void
+t8_write_message_by_dim (const char *message, const T *array, const int dim)
+{
+  std::ostringstream buffer;
+  buffer.precision (6);
+  buffer.setf (std::ios::fixed);
+  buffer << message;
+  for (int i_dim = 0; i_dim < dim; ++i_dim) {
+    buffer << " " << array[i_dim];
+  }
+  t8_debugf ("%s\n", buffer.str ().c_str ());
+}
+
+/* To save an iterate, we will test the ref coords and centroid right in the adapt callback itself. 
+ * Refines as long as the MAXLEVEL is not reached. */
+int
+t8_adapt_callback_with_test (t8_forest_t forest, t8_forest_t forest_from, t8_locidx_t which_tree,
+                             t8_locidx_t lelement_id, t8_eclass_scheme_c *ts, const int is_family,
+                             const int num_elements, t8_element_t *elements[])
+{
+  double tree_ref_coords_by_vertex
+    [3]; /** reference coordinates of the element vertices computed by \ref t8_element_vertex_reference_coords */
+  double tree_ref_coords_by_element_ref_coords
+    [T8_ECLASS_MAX_CORNERS
+     * T8_ECLASS_MAX_DIM]; /** reference coordinates of the element vertices computed by \ref t8_element_reference_coords */
+  double centroid_by_vertices
+    [3]; /** centroid computed via \ref t8_element_vertex_reference_coords -> \ref t8_geometry_evaluate -> mean of all results */
+  double centroid_by_element_ref_coords
+    [3]; /** centroid computed via \ref t8_forest_element_centroid (uses \ref t8_element_reference_coords) -> \ref t8_geometry_evaluate */
+  double batch_coords
+    [T8_ECLASS_MAX_CORNERS
+     * T8_ECLASS_MAX_DIM]; /** reference coordinates of the element vertices computed by \ref t8_get_batch_coords_for_element_type */
+
+  /* compare results of the two different way to obtain tree ref coords */
+  const t8_element_shape_t shape = ts->t8_element_shape (elements[0]);
+  const int num_vertices = t8_eclass_num_vertices[shape];
+  const int elem_dim = t8_eclass_to_dimension[shape];
+  t8_get_batch_coords_for_element_type (shape, batch_coords);
+  t8_debugf ("Testing for shape %s\n", t8_eclass_to_string[shape]);
+  t8_debugf ("with num_vertices %i\n", num_vertices);
+  t8_debugf ("and elem_dim %i\n", elem_dim);
+
+  ts->t8_element_reference_coords (elements[0], batch_coords, num_vertices, tree_ref_coords_by_element_ref_coords);
+  for (int i_vertex = 0; i_vertex < num_vertices; ++i_vertex) {
+    ts->t8_element_vertex_reference_coords (elements[0], i_vertex, tree_ref_coords_by_vertex);
+    t8_debugf ("Checking vertex %i\n", i_vertex);
```

Comment:
Can you add this message via `<<` to the `EXPECT_NEAR`. That way this is printed only for failing tests.

Reply:
I could, but the messages print all coordinates and the `EXPECT_NEAR` only checks one dimension at a time. If we append the message, and multiple coords fail, the same message gets printed multiple times. Moreover, the logs of passing tests is not included in the overall log file, so it should not overcrowd the log files

Reply:
Yes, you're right.